samples/drivers/spi_efr32bg22/src/spi.c

```diff
@@ -0,0 +1,135 @@
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <zephyr/drivers/spi.h>
+
+#define SPI_EFR32BG22_DEV_ID DT_COMPAT_GET_ANY_STATUS_OKAY(spi_efr32bg22)
+
+#define SPI_WORD_SIZE 8
+#define SPI_OP SPI_OP_MODE_MASTER | \
+	       SPI_MODE_CPOL | \
+	       SPI_MODE_CPHA | \
+	       SPI_WORD_SET(SPI_WORD_SIZE) | \
+	       SPI_LINES_SINGLE
+
+#define BUF_SIZE 4
+uint8_t buffer_tx[] = {0x9f, 0x00, 0x00, 0x00};
+uint8_t buffer_rx[BUF_SIZE] = {};
+
+/*
+ * We need 5x(buffer size) + 1 to print a comma-separated list of each
+ * byte in hex, plus a null.
+ */
+uint8_t buffer_print_tx[BUF_SIZE * 5 + 1];
+uint8_t buffer_print_rx[BUF_SIZE * 5 + 1];
+
+struct spi_dt_spec spi_efr32bg22 = SPI_DT_SPEC_GET(SPI_EFR32BG22_DEV_ID, SPI_OP, 0);
+struct gpio_dt_spec spi_enable = GPIO_DT_SPEC_GET(DT_NODELABEL(spi_enable), gpios);
+
+static void to_display_format(const uint8_t *src, size_t size, char *dst)
+{
+	size_t i;
+
+	for (i = 0; i < size; i++) {
+		sprintf(dst + 5 * i, "0x%02x,", src[i]);
+	}
+}
+
+static int spi_read_rdid(struct spi_dt_spec *spec)
+{
+	const struct spi_buf tx_bufs[] = {
+		{
+			.buf = buffer_tx,
+			.len = BUF_SIZE,
+		},
+	};
+	const struct spi_buf rx_bufs[] = {
+		{
+			.buf = buffer_rx,
+			.len = BUF_SIZE,
+		},
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_bufs,
+		.count = ARRAY_SIZE(tx_bufs)
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_bufs,
+		.count = ARRAY_SIZE(rx_bufs)
+	};
+
+	int ret;
+
+	LOG_INF("Start ID readout");
+
+	ret = spi_transceive(spec->bus, &spec->config, &tx, &rx);
+	LOG_INF("SPI transceive finished with code %d", ret);
+	if (ret) {
+		LOG_ERR("Code %d\nSPI transceive failed", ret);
+		return ret;
+	}
+
+	to_display_format(buffer_tx, BUF_SIZE, buffer_print_tx);
+	to_display_format(buffer_rx, BUF_SIZE, buffer_print_rx);
+	LOG_INF("Buffer contents");
+	LOG_INF("tx: %s", buffer_print_tx);
+	LOG_INF("rx: %s", buffer_print_rx);
+
+	LOG_INF("Passed");
+
+	return 0;
+}
+
+static void test_spi(void)
+{
+	LOG_INF("SPI test on buffers TX/RX %p / %p", (void *)buffer_tx, (void *)buffer_rx);
+
+	int ret;
+
+	LOG_INF("Configure SPI Enable GPIO as output");
+	ret = gpio_pin_configure_dt(&spi_enable, GPIO_OUTPUT);
+	if (ret != 0) {
+		LOG_ERR("Failed to configure SPI enable GPIO as output, return code: %d", ret);
+	}
+
+	LOG_INF("Enable SPI bus on board");
+	ret = gpio_pin_set_dt(&spi_enable, 1);
+	if (ret != 0) {
+		LOG_ERR("Failed to enable SPI bus, return code: %d", ret);
+	}
+
+	LOG_INF("Check if SPI is ready");
+	if (!spi_is_ready(&spi_efr32bg22)) {
+		LOG_ERR("SPI device is not ready");
+	}
+
+	LOG_INF("Read ID from MX25R SPI Flash device");
+	if (spi_read_rdid(&spi_efr32bg22)) {
+		return;
+	}
+
+	LOG_INF("Disable SPI bus on board");
+	ret = gpio_pin_set_dt(&spi_enable, 0);
+	if (ret != 0) {
+		LOG_ERR("Failed to disable SPI bus, return code: %d", ret);
+	}
+
+	LOG_INF("Done");
+}
+
+void main(void)
+{
+	test_spi();
+}
```

Comment:
-1 on platform specific samples for interfaces. Also, the SPI enable bit should be part of the SPI driver and be specified in the SPI bindings.

Reply:
Removed the sample. Please note that the SPI enable bit was not used as CS line. It was used for driving on-board switch. The switch connected SPI peripheral used in initial SPI driver tests to the common SPI bus on the platform. This should be handled by users in their own applications so I also removed this part from the PR.
FYI: I'm pasting here a figure from Thunderboard EFR32BG22 User Guide that highlights the usage of SPI enable bit (EFR32_IMU_ENABLE)
![image](https://user-images.githubusercontent.com/33930985/209128754-1320465f-877c-43ea-8fa0-0516e698161b.png)

Reply:
note that on-board switch should either use the `regulator-fixed` compatible, or, if specific to a device you can also use `-gpios` prop to directly toggle the GPIO in the driver (probably combined with PM).

Reply:
I'm not sure if I understand that correctly. Would you like me to reintroduce the SPI enable bit but using existing compatible or the `-gpios` prop? It is specific to one board and one sensor on that board. That is also why I don't think it should be handled by SPI driver. In my opinion this should be handled on application level when user decides to actually power up and use the sensor. Please let me know if I can leave the SPI enable pin removed or if I should rework it.

Reply:
If that's the case, you can define a `fixed-regulator` for that GPIO. Then clients can use the `vin-supply` property to capture the regulator handle and use `regulator_enable/disable` as needed. However, usage of regulators is rarely supported in subsystems, I hope this improves in the future.

Reply:
Added in https://github.com/zephyrproject-rtos/zephyr/pull/53255/commits/7402a6bb6f9fc6a2aa6998f73590f71a957a4c57.